### PR TITLE
Adding custom cursor modifier support

### PR DESCRIPTION
**Pull request by shubhamarora28**
Overview
---
**Issue in existing cusor system:**
Cursor class uses CursorModifier in the following two methods:
- OnActiveModifier(CursorModifier modifier)
- OnPointerSpecificFocusChanged(IPointingSource pointer, GameObject oldFocusedObject, GameObject newFocusedObject)

because of which, if anyone writes their own cursor modifier(implementing ICursorModifier), their cursor modifier code will not be called/executed as OnPointerSpecificFocusChanged will not be able to get the custom cursor modifier component as it is not of the same type as expected by the Cursor class.

Solution:
- Changed OnActiveModifier(ICursorModifier modifier) signature to use ICursorModifier as the input parameter instead of CursorModifier
- Changed OnPointerSpecificFocusChanged default implementation to use ICursorModifier instead of CursorModifier

Implication:
All classes that derive from Cursor abstract class will need to change OnActiveModifier method parameter type

Diagram images of the existing system and proposed changes to the system are attached.

**Current Cursor System:**
cursorsystemdiagram_current

**Proposed changes to cursor system:**
cursorsystemdiagram_proposed

P.S.: Edited the images with MS Paint. Please don't judge. :D
We were not able to generate the class diagram using VS 2017 Professional. Any help on this will be appreciated.

Changes
---
- Modified Cursor.cs to use ICursorModifier instead of CursorModifier
- Modified AnimatedCursor.cs
- Added CustomCursorModifierExample scene
- Added CubeCursor prefab
- Added TestCursor.cs
- Added TestCursorModifier.cs

### CODE REVIEW

**Comment by shubhamarora28** [comment]
@StephenHodgson Brought back the old method and marked it obsolete.

Regarding the test scripts and example scene, MRTK does not have any example of a custom cursor modifier which is why I created the scene. It has three game objects, one without any modifier, one with CursorModifier and one with a custom cursor modifier(TestCursorModifier).

Being a consumer of MRTK, I myself faced some issues while implementing my own cursor related functionality. I think the example scene will be helpful for some people.

File: Assets/HoloToolkit/Input/Scripts/Cursor/Cursor.cs [hunk 1]
```diff
@@ -306,11 +306,29 @@ protected virtual void OnPointerSpecificFocusChanged(IPointingSource pointer, Ga
         /// Override function when a new modifier is found or no modifier is valid
         /// </summary>
         /// <param name="modifier"></param>
+        [System.Obsolete("Use OnActiveModifier(ICursorModifier modifier) instead.")]
         protected virtual void OnActiveModifier(CursorModifier modifier)
         {
             TargetedCursorModifier = modifier;
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Input/Scripts/Cursor/Cursor.cs]: Shouldn't this now be calling the new method with the interface parameter?

File: Assets/HoloToolkit/Input/Scripts/Cursor/AnimatedCursor.cs [hunk 2]
```diff
@@ -69,20 +69,28 @@ public override void OnInputDisabled()
         /// Override to set the cursor animation trigger
         /// </summary>
         /// <param name="modifier"></param>
-        protected override void OnActiveModifier(CursorModifier modifier)
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Input/Scripts/Cursor/AnimatedCursor.cs]: Let's keep the old method and use it to call the method with the interface implementation and mark the old one obsolete.

**Comment by davidkline-ms** [comment]
the july (june18_dev) branch is in stabilization, re-targeting this PR to the htk_development branch

File: Assets/HoloToolkit/Input/Scripts/Cursor/AnimatedCursor.cs (hunk 2 above)
Comment (shubhamarora28) [reply Assets/HoloToolkit/Input/Scripts/Cursor/AnimatedCursor.cs]: In line with the comment I have made above, if we keep the old method, it'll receive null as a parameter when gaze is on the game object with any other implementation of ICursorModifier(e.g., TestCursorModifier) and hence, the new method with the interface implementation will also get null.

File: Assets/HoloToolkit/Input/Scripts/Cursor/Cursor.cs (hunk 1 above)
Comment (shubhamarora28) [reply Assets/HoloToolkit/Input/Scripts/Cursor/Cursor.cs]: Following is the OnPointerSpecificFocusChanged method:
onpointerspecificfocuschanged

Here, if I call the old OnActiveModifier(CursorModifier modifier) method instead of the new one with interface parameter, I need to cast the ICursorModifier object to CursorModifer which will return null in all the cases when newFocusedObject has an implementation of ICursorModifier which is not CursorModifier.
Hence, the old OnActiveModifier(CursorModifier modifier) method will receive null as a parameter and pass null to the new implementation of the method.

In order to avoid this issue, I am calling the new OnActiveModifier with the interface parameter from OnPointerSpecificFocusChanged since ICursorModifier is the parent here and from inside of this method, I am checking if this implementation of ICursorModifier is CursorModifier, only then I call the old OnActiveModifier.
new_onactivemodifier


This keeps the old functionality intact and will not give any errors or produce any unexpected behavior for those users who have custom classes inherited from these.

**Comment by shubhamarora28** [comment]
@StephenHodgson Updated TestCursorModifier to not throw `NotImplementedExceptions`.

File: Assets/HoloToolkit/Boundary/Scripts/BoundaryManager.cs [hunk 3]
```diff
@@ -215,7 +215,7 @@ public Vector3[] TryGetBoundaryRectanglePoints()
             var positions = new Vector3[points2d.Length];
             for (int i = 0; i < points2d.Length; ++i)
             {
-                positions[i] = new Vector3(points2d[i].x, boundaryFloor, points2d[i].y);
+                positions[i] = CameraCache.Main.transform.parent.TransformPoint(new Vector3(points2d[i].x, boundaryFloor, points2d[i].y));
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Boundary/Scripts/BoundaryManager.cs]: This change isn't relevant to custom cursors. Please, either make a new PR and revert this change, or add it to the description.

Comment (shubhamarora28) [reply Assets/HoloToolkit/Boundary/Scripts/BoundaryManager.cs]: This change was made by @keveleigh.

Comment (StephenHodgson) [reply Assets/HoloToolkit/Boundary/Scripts/BoundaryManager.cs]: Oh, I see, you merged from a different branch into this one?

Comment (shubhamarora28) [reply Assets/HoloToolkit/Boundary/Scripts/BoundaryManager.cs]: Yes. I took june18_dev branch as base.

Comment (shubhamarora28) [reply Assets/HoloToolkit/Boundary/Scripts/BoundaryManager.cs]: Sure. Will do that.

Comment (StephenHodgson) [reply Assets/HoloToolkit/Boundary/Scripts/BoundaryManager.cs]: Thanks.

Comment (shubhamarora28) [reply Assets/HoloToolkit/Boundary/Scripts/BoundaryManager.cs]: The changes made in #2518 are already there in my branch.

File: Assets/HoloToolkit/Input/Scripts/Cursor/Cursor.cs (hunk 1 above)
Comment (shubhamarora28) [reply Assets/HoloToolkit/Input/Scripts/Cursor/Cursor.cs]: @StephenHodgson Are we good with this?

**Comment by SimonDarksideJ** [comment]
Can we get an update on this @keveleigh / @StephenHodgson as it's getting stale.

**Comment by wiwei** [comment]
Closing issues older than 180 days. If this is still an issue, please reactivate with recent information.
